tokio-net/src/driver/reactor/sharded_slab/page/slot.rs

```diff
@@ -0,0 +1,105 @@
+use super::super::{Pack, Tid, RESERVED_BITS, WIDTH};
+use super::ScheduledIo;
+use crate::sync::{
+    atomic::{AtomicUsize, Ordering},
+    CausalCell,
+};
+
+#[derive(Debug)]
+pub(crate) struct Slot {
+    gen: AtomicUsize,
+    /// The offset of the next item on the free list.
+    next: CausalCell<usize>,
+    /// The data stored in the slot.
+    item: ScheduledIo,
+}
+
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub(crate) struct Generation {
+    value: usize,
+}
+
+impl Pack for Generation {
+    /// Use all the remaining bits in the word for the generation counter, minus
+    /// any bits reserved by the user.
+    const LEN: usize = (WIDTH - RESERVED_BITS) - Self::SHIFT;
+
+    type Prev = Tid;
+
+    #[inline(always)]
+    fn from_usize(u: usize) -> Self {
+        debug_assert!(u <= Self::BITS);
+        Self::new(u)
+    }
+
+    #[inline(always)]
+    fn as_usize(&self) -> usize {
+        self.value
+    }
+}
+
+impl Generation {
+    fn new(value: usize) -> Self {
+        Self { value }
+    }
+}
+
+impl Slot {
+    pub(super) fn new(next: usize) -> Self {
+        Self {
+            gen: AtomicUsize::new(0),
+            item: ScheduledIo::default(),
+            next: CausalCell::new(next),
+        }
+    }
+
+    #[inline(always)]
+    pub(super) fn get(&self, gen: Generation) -> Option<&ScheduledIo> {
+        let current = self.gen.load(Ordering::SeqCst);
```

Comment:
Thinking about it more, this will load the correct generation, but what is preventing a get concurrent with a remove followed by insert? If there is something, then great. Otherwise, it seems that the only option we have is combining the gen w/ the readiness atomic.